staticDHCPd/doc/customisation/configuration.rst
:tocdepth: 3

Configuration guide
===================
A single config file is the only thing you need to modify in order to have
almost full control over your DHCP server's behaviour.

The file is intended to be largely forwards-compatible, meaning that, when a new
version of *staticDHCPd* is released, you shouldn't have to change anything: the
new engine will massage and inject data as needed to make sure that gaps in your
configuration will be filled with sane defaults, and that old scripting logic
will still work. (Within reason, of course; always check the changelog)

The default file is very minimal in nature. To extend it, simply place the
parameters, described below, that you want to modify into that file and restart
your server.

``conf.py``
-----------
conf.py is, as you may have surmised, a real Python file. It is executed by the
interpreter before *staticDHCPd* is fully online and you can do, with it,
everything that you'd normally expect a Python script to do, including importing
other modules. More information on that is provided in the
:doc:`scripting guide <./scripting>` -- the important thing to note for basic
configurations is that you need to conform to Python syntax when specifying
values:

* ``None``, written capitalised and without quotation, indicates that a value is
  explicitly null.
* Text values must be strings, "like this" or 'like this', with quotation marks.
* Numbers are written naturally, like 6 or 7.32.
* Boolean values are written as ``True`` or ``False``, capitalised, without
  quotation. ``1`` and ``0`` are NOT suitable substitutes.
* Lists are written like ``[1, 2, 'hello']``, and may contain mixed data-types.
* Concatenations of strings may be done using the + operator::
    
    'hello' + 'goodbye' # = 'hellogoodbye'
    #Non-string data, like numbers, must be converted
    'hello' + str(5) # = 'hello5'
    
* Any value defined in the file may be referenced on any later line, like this::
    
    _HELLO = 'hello'
    GREETING = _HELLO + 'goodbye'
    
* Everything is case-sensitive.
 
You may define non-config-parameter values in this file, but it would be a good
idea to prefix their names with an underscore, to avoid potential conflicts with
future upstream changes.

Parameters
----------
The various settings that may be defined within ``conf.py``. These should always
be specified before any custom code you might add, like functions.

General
+++++++
**DEBUG** : boolean : default=False
|||||||||||||||||||||||||||||||||||
* Adds additional information to everything that gets logged
* Does not modify the logging severity levels of individual loggers

**SYSTEM_NAME** : text : default='staticDHCPd'
||||||||||||||||||||||||||||||||||||||||||||||
* The name the system will use to self-identify

**DAEMON** : boolean : default=True
|||||||||||||||||||||||||||||||||||
* Causes the server to daemonise during startup
* Not a good thing to have enabled when doing initial setup because you lose ^C
  and console logs

**PID_FILE** : text, None : default=None
||||||||||||||||||||||||||||||||||||||||
* The path to which a pidfile should be written
* ``'/var/run/staticDHCPd.pid'`` is a good choice

Server
++++++
**UID** : integer : **MUST BE SPECIFIED**
|||||||||||||||||||||||||||||||||||||||||
* The UID under which the server will run, after everything is bound
* This should normally be your system's "nobody"

**GID** : integer : **MUST BE SPECIFIED**
|||||||||||||||||||||||||||||||||||||||||
* The GID under which the server will run, after everything is bound
* This should normally be your system's "nobody"

**DHCP_SERVER_IP** : text : **MUST BE SPECIFIED**
|||||||||||||||||||||||||||||||||||||||||||||||||
* The IP of the interface to use for DHCP traffic
* This value must be a specific IP address: ``'0.0.0.0'`` is invalid

**DHCP_RESPONSE_INTERFACE** : text : default='-'
||||||||||||||||||||||||||||||||||||||||||||||||
* The value is the lexical name of the interface from which responses
  should be sent, like ``'eth0'``
* A special value of ``'-'`` may be used to automatically resolve the interface
  based on `DHCP_SERVER_IP`
* If set, response-packets will be crafted from layer 2, allowing for unicast
  OFFERs in response to DISCOVERs, if the client set the broadcast bit
* For the vast majority of environments where staticDHCPd is likely to be used,
  it is perfectly fine, and recommended, to disable this feature with ``None``
  
  * This feature was historically not enabled by default because it incurs a
    slight performance penalty (calculating IP and UDP checksums in Python) and
    some additional overhead (interfacing with *libpcap* from Python if Linux's
    *AF_PACKET* (which has its own syscalls) is not available) on all
    broadcast-scope exchanges, whether they use unicast MAC-broadcasts or not,
    plus a persistent filehandle and some memory, and complications with VLANs
    if the server is expected to handle them
  * Unfortunately, some Windows firewalling configurations require it
  * Surprisingly, the Linux kernel's DHCP autoconfiguration mechanism does, too

**DHCP_RESPONSE_INTERFACE_QTAGS** : list : default=None
|||||||||||||||||||||||||||||||||||||||||||||||||||||||
* If **DHCP_RESPONSE_INTERFACE** is set and qtags are required on that
  interface, they may be supplied in order of appearance (head to tail) as TCI
  blocks
* Format::

    (Priority Control Point (0-7),
     Drop Eligible Indicator (True/False),
     VLAN Identifier (1-4094),)

* Example: ``[(3, True, 42), (1, False, 77)]`` -> VLAN 42 with nested VLAN 77

**DHCP_SERVER_PORT** : integer : default=67
|||||||||||||||||||||||||||||||||||||||||||
* The port on which to listen for DHCP queries

**DHCP_CLIENT_PORT** : integer : default=68
|||||||||||||||||||||||||||||||||||||||||||
* The port on which to respond to DHCP clients

**PROXY_PORT** : integer, None : default=None
|||||||||||||||||||||||||||||||||||||||||||||
* The port on which to listen for ProxyDHCP requests, often referred to as PXE,
  if required
* ProxyDHCP normally runs on ``4011``

* This option used to be named **PXE_PORT**, which will still be honoured.

**RELAY_PORT_QUIRKS** : dict[str:int], None : default=None
||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
* A dictionary of DHCP types to port numbers, used to override the destination
  port when sending a response to a DHCP relay.
* Ideally, this should never be needed, since all relay destination ports are
  supposed to be ``DHCP_SERVER_PORT``, with the relay being responsible for
  flipping that to ``DHCP_CLIENT_PORT`` for ``OFFER`` responses, but some relays
  that may be BOOTP-literal just pass the response through exactly as received.
* Example: ``{"DHCP_OFFER": 68}``
* Relevant strings:
  * ``DHCP_OFFER``
  * ``DHCP_ACK``
  * ``DHCP_NACK``
  * ``DHCP_FORCERENEW``
  * ``DHCP_LEASEUNASSIGNED``
  * ``DHCP_LEASEUNKNOWN``
  * ``DHCP_LEASEACTIVE``

Caching
+++++++
Caches are only enabled if the database engine supports them and **USE_CACHE**
is set.

**CACHING_MODEL** : text : default="in-process"
|||||||||||||||||||||||||||||||||||||||||||||||
* The database-engine to use for handling static lease data

  * ``'in-process'``: manage the cache using in-process constructs, either
    memory or an on-disk file accessed exclusively by *staticDHCPd*
  * ``'memcached'``: use a *memcached* server as an external store

**DISK_CACHE** : boolean : default=False
|||||||||||||||||||||||||||||||||||||||||||
* Causes the local cache to be managed as a local file, rather than a purely
  in-memory construct

  * For largeish caches that are accessed infrequently, but where local access
    through disk (or hotspot-cached pages in RAM) is still expected to be more
    efficient than a remote database call
  * This file will be temporary, unless **DISK_CACHE_PERSISTENT** is set; in that
    case, the file will be the same
    
* This option used to be named **CACHE_ON_DISK**, which will still be honoured.

**DISK_CACHE_PERSISTENT** : text : default=None
||||||||||||||||||||||||||||||||||||||||||
* Causes the cache to be written to a local database file, which will be used
  when *staticDHCPd* is restarted, to provide durability against unstable
  databases
* The value of this option is the path to the file;
  ``'/var/tmp/staticDHCPd.db'`` is usually a good choice
* If **DISK_CACHE** is set, this file will be used directly as a buffer; if
  not, the contents of this file will be read into memory and this file will be
  write-only at runtime

* This option used to be named **PERSISTENT_CACHE**, which will still be honoured.
  
**MEMCACHED_HOST** : text : default=None
||||||||||||||||||||||||||||||||||||||||
* The address of the *memcached* server to use with the corresponding model

**MEMCACHED_PORT** : integer : default=11211
||||||||||||||||||||||||||||||||||||||||||||
* The port of the *memcached* server to use with the corresponding model

**MEMCACHED_AGE_TIME** : integer : default=300
||||||||||||||||||||||||||||||||||||||||||||||
* The lifetime to apply to cached data within the *memcached* server

Database
++++++++
**DATABASE_ENGINE** : text, None : **MUST BE SPECIFIED**
||||||||||||||||||||||||||||||||||||||||||||||||||||||||
* The database-engine to use for handling static lease data

  * One of ``None``, ``'SQLite'``, ``'PostgreSQL'``, ``'Oracle'``, ``'MySQL'``,
    ``'INI'``
  
* Only parameters related to the chosen database engine need to be supplied
* You may alternatively supply a callable that provides an instance of
  :class:`databases.generic.Database`, if you want to implement your own engine
  without messing with core code
  
  * The callable must require no arguments
  * To keep ``conf.py`` clean, define your engine in a separate module, like
    ``extensions/_mydb.py`` and import it just before **DATABASE_ENGINE** is
    set, referencing a callable declared therein
  * If you need to configure it differently for each server, declare a lambda
    or short function in ``conf.py`` and make changes to that as needed
  * It is sane to inherit from subclasses like
    :class:`databases.generic.CachingDatabase`; it is probably safe to inherit
    from the :class:`databases._sql._SQLDatabase` family, too, but its internal
    implementation is technically private
  * If you need to tie into :ref:`callbacks <scripting-callbacks>`, like
    reinitialisation, you should do this as part of the callable's logic; the
    ``callbacks`` object is not available at the time that ``conf.py`` is first
    processed, but it is available while the callable is executed
    
    * Alternatively, you can create the object at the ``conf.py`` level, write a
      simple ``lambda : _MY_INSTANTIATED_DATABASE_OBJECT`` as the callable, and
      operate on ``_MY_INSTANTIATED_DATABASE_OBJECT`` in ``init()``, for
      consistency
    * Note: the database object's ``reinitialise()`` method is automatically
      registered, so you only need to tie into callbacks if you need behaviour
      that cannot be captured there
      
Database:None
|||||||||||||
No parameters to set. This database is only useful if you are exclusively using
``handleUnknownMAC()``, described in the :doc:`scripting guide <./scripting>`,
to provision addresses.

Database:SQLite
|||||||||||||||
**CASE_INSENSITIVE_MACS** : boolean : default=False
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
* Forces case-insensitive comparisons for MACs
* If this matters to you, you should create a NOCASE index over `maps:mac`
  instead, for greater efficiency

**USE_CACHE** : boolean : default=False
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
* Causes data retrieved from the database to be stored in memory until the
  cache is flushed via reinitialisation
* For SQLite, this should normally be ``False``

**EXTRA_MAPS** : list : default=None
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
* Any non-standard fields to read from the `maps` table, which will be
  provided in ``definition.extra``, keyed as `maps.$COLUMN`

**EXTRA_SUBNETS** : list : default=None
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
* Any non-standard fields to read from the `subnets` table, which will be
  provided in ``definition.extra``, keyed as `subnets.$COLUMN`

**SQLITE_FILE** : text : *MUST BE SPECIFIED if using SQLite*
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
* The path to the file that contains your SQLite database

Database:PostgreSQL
|||||||||||||||||||
**CASE_INSENSITIVE_MACS** : boolean : default=False
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
* Forces case-insensitive comparisons for MACs
* If this matters to you, you should create a lower() index over `maps:mac`
  instead, for greater efficiency

**USE_CACHE** : boolean : default=False
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
* Causes data retrieved from the database to be stored in memory until the
  cache is flushed via reinitialisation
* Can greatly improve performance in stable, high-load environments

**EXTRA_MAPS** : list : default=None
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
* Any non-standard fields to read from the `maps` table, which will be
  provided in ``definition.extra``, keyed as `maps.$COLUMN`

**EXTRA_SUBNETS** : list : default=None
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
* Any non-standard fields to read from the `subnets` table, which will be
  provided in ``definition.extra``, keyed as `subnets.$COLUMN`

**USE_POOL** : boolean : default=True
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
* Causes database connections to pull from a pool by default, reducing

CHANGELOG
This is legacy information, but it is not documented elsewhere.

---

# 1.6.4 #
  * Backported a feature from 2.0.0 that lets users define their own custom database engines without needing to touch the core code.

---

# 1.5.8, 1.6.3 #
  * Fixed a long-standing issue in which PXE clients that sent broadcast packets containing `ciaddr` were responded to via broadcast, rather than unicast to that `ciaddr` value.

---

# 1.6.2 : staticDHCPd #
  * Refactored the database component, breaking coupling to SQL
  * Implemented support for using an INI file as an alternative to an SQL-based database, borrowing from contributions by John Stowers

---

# 1.6.1 : staticDHCPd #
  * Massive re-architecture of the directory structures
    * staticDHCPd and libpydhcpserver now separated
    * conf/ directory added for source-based testing and deployment
    * /etc/staticDHCPd/ now recognised as a configuration path
  * Better deployment and operation logic
    * Setup scripts for both components of the system
    * Basic unified installer
      * Upgrade-friendly
    * Platform auto-detection and initscript-setup-help
    * No more need for hacks to access custom modules imported relative to `conf.py`
  * Fixed a tab-based indentation issue in 1.6.0 (thanks to John Stowers for pointing it out)
  * `main.py` is now `staticDHCPd`

---

# 1.6.0 (svn-only) #
  * Support for dynamic provisioning.
  * Support for true daemon operation. (Not strictly necessary, but I just realised I somehow left it out, which is odd because it's in everything else I've written)

---

# 1.5.7 #
  * PXE will now work even when ALLOW\_LOCAL\_DHCP is False. This failed before because PXE is strictly unicast, so it doesn't use a relay, even when the other half of the negotiation does.

---

# 1.5.6 #
  * Fixes a significant issue that prevented special DHCP options from being parsed, thanks to Aditya Kulkarni.
  * Fixes an issue that, despite what 1.5.3 said, prevented PXE from working properly, since the socket was never checked for pending data.
  * Fixes the lack of e-mail timeouts and allows ports to be specified.

---

# 1.5.5 #
  * Primarily a bugfix release, and one long overdue.
    * Addresses a number of issues affecting REBIND behaviour and errant NAKs ([issue 7](https://code.google.com/p/staticdhcpd/issues/detail?id=7)).
    * Adds Python 2.5 backwards-compatibility thanks to a patch from Mark Schloesser.
    * Fixes an error-logging issue that caused still more errors to be thrown, thanks again to Mark ([issue 8](https://code.google.com/p/staticdhcpd/issues/detail?id=8)).
    * Fixes issues related to in-code typos that prevent custom attributes from being set correctly in DHCP packets ([issue 10](https://code.google.com/p/staticdhcpd/issues/detail?id=10)).
  * Added some diagnosis logic and a strToPaddedList() function based on patches from Andrew Xuan.

---

# 1.5.4 #
  * Made it easy to upgrade by introducing a buffering setting-default system
  * Refactored some code out of places it didn't need to be
  * Severed a reverse-dependency between libpydhcpserver and staticDHCPd

---

# 1.5.3 (svn-only) #
  * Vendor options are automatically stripped from received packets and exposed to users through a `vendor` parameter
  * Support for a PXE port has been added; see the FAQ for more details
    * Addresses an issue reported by Andrew Xuan
  * Added support for easy logging from user-defined code

---

# 1.5.2 (svn-only) #
  * Addressed an issue with Cisco relays not accepting packets from sockets other than 67
  * Added support for Oracle

---

# 1.5.1 #
  * Fixed a bug with Postgres's configuration, preventing it from being used unless MySQL was configured, too
  * Fixed a documentation error in the README: no netmask was defined

---

# 1.5.0 #
  * Database-flow overhaul
    * Added support for Postgres
    * Added connection pooling
    * Removed concurrency support from SQLite, since it was redundant
  * Converted all source to space-based indentation (you'll want to whitespace-ignorant diff your old <tt>conf.py</tt> files)
  * Cleaned up webserver logging a little

---

# 1.5.0 (preview) #
  * Refactored pydhcplib into libpydhcpserver
    * All code thoroughly commented.
  * Added support for [RFC3442](http://www.rfc-archive.org/getrfc.php?rfc=3442) (option 121)
    * The data to be sent must be crafted by hand.
    * If there is any demand, a special class will be created.
  * Added support for [RFC3495](http://www.rfc-archive.org/getrfc.php?rfc=3495) (option 122)
    * The data to be sent must be crafted by hand.
  * Added support for [RFC3825](http://www.rfc-archive.org/getrfc.php?rfc=3825) (option 123)
    * The data to be sent must be crafted by hand.
    * A special class will be handled to ease this process in a later version.
  * Added support for [RFC3925](http://www.rfc-archive.org/getrfc.php?rfc=3925) (options 124 and 125)
    * The data to be sent must be crafted by hand.
  * Added support for [RFC4039](http://www.rfc-archive.org/getrfc.php?rfc=4039) (option 80)
  * Added support for [RFC4280](http://www.rfc-archive.org/getrfc.php?rfc=4280) (options 88 and 89)
  * Added support for [RFC4578](http://www.rfc-archive.org/getrfc.php?rfc=4578) (options 93, 94, and 97)
  * Added support for [RFC4776](http://www.rfc-archive.org/getrfc.php?rfc=4776) (option 99)
    * The data must be crafted by hand, due to the complexity of the option.
    * If there is sufficient demand, a special class may be created.
  * Added support for [RFC4833](http://www.rfc-archive.org/getrfc.php?rfc=4833) (options 100 and 101)
  * Added support for [RFC5071](http://www.rfc-archive.org/getrfc.php?rfc=5071) (options 208-211)
  * Added support for [RFC5192](http://www.rfc-archive.org/getrfc.php?rfc=5192) (option 136)
  * Added support for [RFC5223](http://www.rfc-archive.org/getrfc.php?rfc=5223) (option 137)
  * Added support for [RFC5417](http://www.rfc-archive.org/getrfc.php?rfc=5417) (option 138)

---

# 1.4.1 #

staticDHCPd/staticdhcpdlib/config.py
# -*- encoding: utf-8 -*-
"""
staticdhcpdlib.config
=====================
Provides a buffer to seed options with default values to make upgrading easier
for end users who do not need to manage any newly added features.

Also handles the process of determining where config values should be accessed.

WARNING
-------
If you are attempting to customise your environment, edit conf.py instead.
If testing, it will likely be in conf/; if installed, in /etc/staticDHCPd/;
or, if upgrading from an older version, in the same directory as main.py.

This file is intended for internal use only and modifications here will
probably lead to headaches later.

Legal
-----
This file is part of staticDHCPd.
staticDHCPd is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.

(C) Neil Tallim, 2021 <neil.tallim@linux.com>
"""
#Get the "conf" module from somewhere
conf = None
import os
import sys
from importlib.machinery import SourceFileLoader

if 'STATICDHCPD_CONF_PATH' in os.environ:
    conf_search_paths = [os.path.dirname(os.environ['STATICDHCPD_CONF_PATH'])]
else:
    conf_search_paths = [os.path.join(os.getcwd(), 'conf'), '/etc/staticDHCPd']

for conf_path in conf_search_paths:
    extensions_path = os.path.join(conf_path, 'staticDHCPd_extensions')
    sys.path.append(conf_path)
    try: #Attempt to import conf.py from the path
        conf = SourceFileLoader('conf', os.path.join(conf_path, 'conf.py')).load_module()
        print(repr(conf))
        print(dir(conf))
    except IOError:
        sys.path.remove(conf_path)
    else:
        break
else:
    raise ImportError("Unable to find a suitable instance of conf.py; searched: {!r}".format(conf_search_paths))

del conf_search_paths
del conf_path
del os
del sys
del SourceFileLoader

#Options passed through from conf.py
#For explanations, please consult that file.
##############################################################################
_defaults = {}

#General settings
#######################################
_defaults.update({
    'DEBUG': False,
    'DAEMON': True,
    'SYSTEM_NAME': 'staticDHCPd',
    'PID_FILE': None,
})

#Server settings
#######################################
_defaults.update({
    'DHCP_RESPONSE_INTERFACE': '-',
    'DHCP_RESPONSE_INTERFACE_QTAGS': None,
    'DHCP_SERVER_PORT': 67,
    'DHCP_CLIENT_PORT': 68,
    'PROXY_PORT': None,
    'RELAY_PORT_QUIRKS': None,
})

#Database settings
#######################################
_defaults.update({
    'USE_CACHE': False,
    'CACHING_MODEL': 'in-process',

    'DISK_CACHE': False,
    'DISK_CACHE_PERSISTENT': None,

    'MEMCACHED_HOST': None,
    'MEMCACHED_PORT': 11211,
    'MEMCACHED_AGE_TIME': 300, #5 minutes

    'CASE_INSENSITIVE_MACS': False,

    'EXTRA_MAPS': None,
    'EXTRA_SUBNETS': None,

    'USE_POOL': True,

    'POSTGRESQL_HOST': None,
    'POSTGRESQL_PORT': 5432,
    'POSTGRESQL_SSLMODE': 'disable',
    'POSTGRESQL_MAXIMUM_CONNECTIONS': 4,

    'ORACLE_MAXIMUM_CONNECTIONS': 4,

    'MYSQL_HOST': None,
    'MYSQL_PORT': 3306,
    'MYSQL_MAXIMUM_CONNECTIONS': 4,
})

#Server behaviour settings
#######################################
_defaults.update({
    'ALLOW_LOCAL_DHCP': True,
    'ALLOW_DHCP_RELAYS': False,
    'ALLOWED_DHCP_RELAYS': [],
    'ENABLE_RAPIDCOMMIT': True,

    'AUTHORITATIVE': False,
    'NAK_RENEWALS': False,

    'UNAUTHORIZED_CLIENT_TIMEOUT': 60,
    'MISBEHAVING_CLIENT_TIMEOUT': 150,
    'ENABLE_SUSPEND': True,
    'SUSPEND_THRESHOLD': 10,
})

#Logging settings
#######################################
_defaults.update({
    'LOG_FILE': None,
    'LOG_FILE_HISTORY': 7,
    'LOG_FILE_SEVERITY': 'WARN',
    'LOG_CONSOLE_SEVERITY': 'INFO',
})

#Webservice settings
#######################################
_defaults.update({
    'WEB_ENABLED': True,
    'WEB_IP': '0.0.0.0',
    'WEB_PORT': 30880,
    'WEB_LOG_HISTORY': 200,
    'WEB_LOG_SEVERITY': 'INFO',
    'WEB_LOG_MAX_HEIGHT': 400,
    'WEB_DIGEST_USERNAME': None,
    'WEB_DIGEST_PASSWORD': None,
    'WEB_DASHBOARD_SECURE': False,
    'WEB_REINITIALISE_CONFIRM': True,
    'WEB_REINITIALISE_SECURE': False,
    'WEB_REINITIALISE_HIDDEN': False,
    'WEB_REINITIALISE_ENABLED': True,
    'WEB_DASHBOARD_ORDER_LOG': 1000,
    'WEB_HEADER_TITLE': True,
    'WEB_HEADER_CSS': True,
    'WEB_HEADER_FAVICON': True,
})

#E-mail settings
#######################################
_defaults.update({
    'EMAIL_ENABLED': False,
    'EMAIL_PORT': 25,
    'EMAIL_TIMEOUT': 4.0,
    'EMAIL_SUBJECT': "staticDHCPd encountered a problem",
    'EMAIL_USER': None,
})


#Construct a unified namespace
#######################################
_REMAPPED_KEYS = {
    #caching keys weren't very consistent and other caching models started to get introduced
    'CACHE_ON_DISK': 'DISK_CACHE',
    'PERSISTENT_CACHE': 'DISK_CACHE_PERSISTENT',
    
    #PXE_PORT was renamed to PROXY_PORT because its role was misunderstood
    'PXE_PORT': 'PROXY_PORT',
} #keys that have had name-changes since prior versions
for key in [k for k in dir(conf) if k.isupper()]: #Copy everything that looks like a constant.
    new_key = _REMAPPED_KEYS.get(key, key)
    globals()[new_key] = getattr(conf, key)
del _REMAPPED_KEYS

for (key, value) in _defaults.items():
    if not key in globals():
        globals()[key] = value
del _defaults

#Bind known functions and handle backwards-compatibility
#######################################
import inspect
if hasattr(conf, 'init'):
    init = conf.init
else:
    init = lambda *args, **kwargs : None
if hasattr(conf, 'filterPacket'):
    filterPacket = conf.filterPacket
else:
    filterPacket = lambda *args, **kwargs : True

if hasattr(conf, 'handleUnknownMAC'):
    if inspect.getfullargspec(conf.handleUnknownMAC).args == ['mac']:
        #It's pre-2.0.0, so wrap it for backwards-compatibility
        from .databases.generic import Definition
        def handleUnknownMAC(packet, method, mac, client_ip, relay_ip, port):
            result = conf.handleUnknownMAC(mac)
            if result is not None:
                return Definition(
                    ip=result[0], lease_time=result[7],
                    subnet=result[8], serial=result[9],
                    hostname=result[1],
                    subnet_mask=result[2],
                    broadcast_address=result[3],
                    domain_name=result[4],
                    domain_name_servers=result[5],
                    ntp_servers=result[6],
                )
            return None
    else:
        handleUnknownMAC = conf.handleUnknownMAC
else:
    handleUnknownMAC = lambda *args, **kwargs : None

if hasattr(conf, 'filterRetrievedDefinitions'):
    filterRetrievedDefinitions = conf.filterRetrievedDefinitions
else:
    def filterRetrievedDefinitions(definitions, *args, **kwargs):
        raise ValueError('No handler exists for multi-definition matches; implement filterRetrievedDefinitions()')

if hasattr(conf, 'loadDHCPPacket'):
    if inspect.getfullargspec(conf.loadDHCPPacket).args == ['packet', 'mac', 'client_ip', 'relay_ip', 'subnet', 'serial', 'pxe', 'vendor']:
        #It's pre-2.0.0, so wrap it for backwards-compatibility
        import collections
        __PXEOptions = collections.namedtuple("PXEOptions", (
            'client_system',
            'client_ndi',
            'uuid_guid',
        ))
        del collections
        
        def loadDHCPPacket(packet, method, mac, definition, relay_ip, port, source_packet):
            vendor_class = None
            vendor_specific = None
            if source_packet.isOption('vendor_class'):
                vendor_class = tuple(sorted(source_packet.getOption('vendor_class', convert=True).items()))
            if source_packet.isOption('vendor_specific'):
                vendor_specific = tuple((k, tuple(sorted(v.items()))) for (k, v) in sorted(source_packet.getOption('vendor_specific', convert=True).items()))
                
            pxe_options = None
            if port == PROXY_PORT:
                option_93 = source_packet.getOption(93, convert=True) #client_system
                option_94 = source_packet.getOption(94) #client_ndi
                option_97 = source_packet.getOption(97) #uuid_guid
                pxe_options = __PXEOptions(
                    option_93,
                    option_94 and tuple(option_94),
                    option_97 and (option_97[0], option_97[1:]),
                )
                
            return conf.loadDHCPPacket(
                packet, mac, definition.ip, relay_ip, definition.subnet, definition.serial,
                pxe_options,
                (
                    source_packet.getOption('vendor_specific_information'),
                    source_packet.getOption('vendor_class_identifier', convert=True),
                    vendor_class,
                    vendor_specific,
                ),
            )
    else:
        loadDHCPPacket = conf.loadDHCPPacket
else:
    loadDHCPPacket = lambda *args, **kwargs : True
del inspect

#Inject namespace elements into conf.
##############################################################################
import libpydhcpserver.dhcp_types.conversion as conversion
conf.listToIP = conversion.listToIP
conf.listToIPs = conversion.listToIPs
conf.ipToList = conversion.ipToList
conf.ipsToList = conversion.ipsToList
conf.listToInt = conversion.listToInt
conf.listToInts = conversion.listToInts
conf.listToLong = conversion.listToLong
conf.listToLongs = conversion.listToLongs
conf.intToList = conversion.intToList
conf.intsToList = conversion.intsToList
conf.longToList = conversion.longToList
conf.longsToList = conversion.longsToList
conf.strToList = conversion.strToList
conf.strToPaddedList = conversion.strToPaddedList
conf.listToStr = conversion.listToStr
del conversion

import libpydhcpserver.dhcp_types.rfc as rfc
conf.rfc3046_decode = rfc.rfc3046_decode
conf.rfc3925_decode = rfc.rfc3925_decode
conf.rfc3925_125_decode = rfc.rfc3925_125_decode
conf.rfc1035_plus = rfc.rfc1035_plus
conf.rfc2610_78 = rfc.rfc2610_78
conf.rfc2610_79 = rfc.rfc2610_79
conf.rfc3361_120 = rfc.rfc3361_120
conf.rfc3397_119 = rfc.rfc3397_119
conf.rfc3442_121 = rfc.rfc3442_121
conf.rfc3925_124 = rfc.rfc3925_124
conf.rfc3925_125 = rfc.rfc3925_125
conf.rfc4174_83 = rfc.rfc4174_83
conf.rfc4280_88 = rfc.rfc4280_88
conf.rfc5223_137 = rfc.rfc5223_137
conf.rfc5678_139 = rfc.rfc5678_139
conf.rfc5678_140 = rfc.rfc5678_140
del rfc

import logging
logger = logging.getLogger('conf')
conf.writeLog = logger.warning
conf.logger = logger
del logger
del logging

from . import system
from . import statistics
from . import web
class callbacks(object):
    """
    A data-namespace, used to isolate callback-management functions.
    """
    systemAddReinitHandler = staticmethod(system.registerReinitialisationCallback)
    systemRemoveReinitHandler = staticmethod(system.unregisterReinitialisationCallback)
    systemAddTickHandler = staticmethod(system.registerTickCallback)
    systemRemoveTickHandler = system.unregisterTickCallback

    statsAddHandler = staticmethod(statistics.registerStatsCallback)
    statsRemoveHandler = statistics.unregisterStatsCallback

    WEB_METHOD_DASHBOARD = web.WEB_METHOD_DASHBOARD
    WEB_METHOD_TEMPLATE = web.WEB_METHOD_TEMPLATE
    WEB_METHOD_RAW = web.WEB_METHOD_RAW
    webAddHeader = staticmethod(web.registerHeaderCallback)
    webRemoveHeader = staticmethod(web.unregisterHeaderCallback)
    webAddDashboard = staticmethod(web.registerDashboardCallback)
    webRemoveDashboard = staticmethod(web.unregisterDashboardCallback)
    webAddMethod = staticmethod(web.registerMethodCallback)
    webRemoveMethod = staticmethod(web.unregisterMethodCallback)
del system
del statistics
del web
conf.callbacks = callbacks

class _Namespace(object):
    """
    A data-namespace, used to centralise extensions-configuration values.
    """
    __final = False #: If True, then no new layers will be created

    def __init__(self, final=False):
        """
        :param bool final: ``False`` if new namespaces may be automatically
                           created beneath this one.
        """
        self.__final = final

    def __enter__(self):
        return self

    def __exit__(self, type, value, tb):
        return False

    def __getattr__(self, name):
        if name.startswith('__'):
            return object.__getattr__(self, name)

        if self.__final:
            raise AttributeError("Namespace does not contain '{}'".format(name))
        namespace = self.__class__(final=True)
        object.__setattr__(self, name, namespace)
        return namespace

    def extension_config_iter(self):
        """
        Produces an iterable object that enumerates all interesting elements in
        the namespace.

        :return: An iterable object that generates ``(key, value)`` tuples.
        """
        for key in [k for k in dir(self) if not k.startswith('_') and not k.startswith('extension_config_')]: #Copy everything that looks useful
            yield (key, getattr(self, key))

    def extension_config_dict(self):
        """
        Produces a dictionary containing all interesting elements in the
        namespace.

        :return dict: User-set elements in the namespace.
        """
        return dict(self.extension_config_iter())

    def extension_config_merge(self, defaults, required):
        """
        Creates a namespace model from `defaults` before overlaying anything
        defined in this namespace instance, then ensuring that all required
        attributes exist.

        Normal usage will be something like the following::

            CONFIG = this_object.extension_config_merge(defaults={
                'DEFAULT_THING': 5, #Your description of this field
            }, required=[
                'REQUIRED_THING', #Your description (with typing) of what this field means
            ])

        This is effectively self-documenting and it guarantees you'll have
        access to every attribute you want.

        :param dict defaults: The default attributes for the namespace, if not
                              already present.
        :param collection required: A collection of required attribute names.
        :return dict: `defaults` augmented with elements defined in the
                      namespace, then validated to ensure all ``required``
                      elements are present in some form.
        :raise AttributeError: A required attribute is missing.
        :raise ValueError: The namespace cannot be merged.
        """
        if not self.__final:
            raise ValueError("Unable to merge a non-final namespace")

        namespace = defaults.copy()
        namespace.update(self.extension_config_iter())
        for key in required:
            if not key in namespace:
                raise AttributeError("Merged result does not contain '{}'".format(key))
        return namespace
conf.extensions = _Namespace()
